README.md

```diff
@@ -35,8 +35,107 @@ Network configuration
-provides facilities to attach a tap device to the Firecracker VM, but
-the client is responsible for further configuration.
+provides facilities to:
+* Attach a pre-created tap device, optionally with static IP configuration, to
+  the VM. This is referred to as a "static network interface".
+* Create a tap device via [CNI](https://github.com/containernetworking/cni) plugins, 
+  which will then be attached to the VM automatically by the SDK. This is referred 
+  to as a "CNI network interface"
+  
+### CNI
+If a VM is configured with a CNI network interface, by default CNI configuration
+files will be sought from `/etc/cni/conf.d` and CNI plugins will be sought under
+`/opt/cni/bin` (both of these values can be overridden via API fields).
+
+It's currently highly recommended to use CNI configuration that includes
+[tc-redirect-tap](cni/Makefile) as a chained plugin. This will allow you to
+adapt pre-existing CNI plugins/configuration to a tap device usable by a
+Firecracker VM.
+
+#### Example
+
+With the following file at `/etc/cni/conf.d/fcnet.conflist`:
+```
+{
+  "name": "fcnet",
+  "cniVersion": "0.3.1",
+  "plugins"": [
+    {
+      "type": "ptp",
+      "ipMasq": true,
+      "ipam": {
+        "type": "host-local",
+        "subnet": "192.168.127.0/24",
+        "resolvConf": "/etc/resolv.conf"
+      }
+    },
+    {
+      "type": "tc-redirect-tap"
+    }
+  ]
+}
+```
+
+and the 
+[`ptp`](https://github.com/containernetworking/plugins/tree/master/plugins/main/ptp), 
+[`host-local`](https://github.com/containernetworking/plugins/tree/master/plugins/ipam/host-local) 
+and [`tc-redirect-tap`](cni/Makefile)
+CNI plugin binaries installed under `/opt/cni/bin`, you can specify, in the Go SDK API, 
+a `Machine` with the following `NetworkInterface`:
+```
+{
+  NetworkInterfaces: []firecracker.NetworkInterface{{
+    CNIConfiguration: &firecracker.CNIConfiguration{
+      NetworkName: "fcnet",
+      IfName: "veth0",
+    },
+  }}
+}
+```
+
+Note that `NetworkName` in the `CNIConfiguration` of the API matches the `name` field 
+specified inside the `/etc/cni/conf.d/fcnet.conflist` file.
+
+With the above configuration, when the Firecracker VM is started the SDK will invoke
+CNI and place the final VM inside the resultant network namespace. The end result being:
+* Outside the network namespace, a single veth endpoint created by the `ptp` plugin will
+  exist with a static IP from the `host-local` plugin (i.e. `192.168.127.1`)
+* Inside the VM's network namespace:
+    * The other side of the veth device will exist with name `veth0`, as specified by the
+      `IfName` parameter above, and a different IP (i.e. `192.168.127.2`)
+    * The tap device created by `tc-redirect-tap`, which will not have an IP but will have
+      all of its traffic mirrored with the `veth0` device
+* Inside the actual Firecracker VM guest:
+    * A network interface with the same IP as that of `veth0` (i.e. `192.168.127.2`)
```

Comment:
What's the interface name inside the VM guest?

Reply:
In practice it's often `eth0`, but it's gonna be up to the VM guest's OS AFAIK. It's not something we configure or need to know at the moment. We just supply the `ip=...` boot params and then the kernel applies those to the first non-loopback interface it finds (which is part of why we validate that if IP configuration is supplied, only up to one network interface can be provided).

In the future, if we move away from the `ip=...` implementation, we may need to start caring about this and/or having the user provide values for guest interface names, but until that time I don't think it's worth mentioning.

Reply:
It would be good to address that point in the bullets here

Reply:
I'll add `the name of the interface is determined by the Guest OS, not the SDK`